tools/nightly_checkout.py

```diff
@@ -0,0 +1,252 @@
+def _get_listing_osx(source_dir):
+    # oddly, these are .so files even on Mac
+    listing = glob.glob(os.path.join(source_dir, "*.so"))
+    listing.append(os.path.join(source_dir, "version.py"))
+    listing.extend(glob.glob(os.path.join(source_dir, "lib", "*.dylib")))
+    listing.append(os.path.join(source_dir, "bin"))
+    return listing
+
+
+def _get_listing_win(source_dir):
+    listing = glob.glob(os.path.join(source_dir, "*.pyd"))
+    listing.append(os.path.join(source_dir, "version.py"))
+    listing.extend(glob.glob(os.path.join(source_dir, "lib", "*.lib")))
+    listing.extend(glob.glob(os.path.join(source_dir, "lib", "*.dll")))
+    listing.append(os.path.join(source_dir, "bin"))
+    return listing
+
+
+def _get_listing(source_dir, platform):
+    if platform.startswith("linux"):
+        listing = _get_listing_linux(source_dir)
+    elif platform.startswith("osx"):
+        listing = _get_listing_osx(source_dir)
+    elif platform.startswith("win"):
+        listing = _get_listing_win(source_dir)
+    else:
+        raise RuntimeError(f"Platform {platform!r} not recognized")
+    return listing
+
+
+def _remove_existing(trg, is_dir):
+    if os.path.exists(trg):
+        if is_dir:
+            shutil.rmtree(trg)
+        else:
+            os.remove(trg)
+
+def _move_single(src, target_dir, mover, verb):
+    is_dir = os.path.isdir(src)
+    base = os.path.basename(src)
+    trg = os.path.join(target_dir, base)
+    _remove_existing(trg, is_dir)
+    # move over new files
+    if is_dir:
+        os.makedirs(trg, exist_ok=True)
+        for root, dirs, files in os.walk(src):
+            relroot = root[len(src):]
+            for name in files:
+                relname = os.path.join(relroot, name)
+                s = os.path.join(src, relname)
+                t = os.path.join(trg, relname)
+                print(f"{verb} {s} -> {t}")
+                mover(s, t)
+            for name in dirs:
+                relname = os.path.join(relroot, name)
+                os.makedirs(os.path.join(trg, relname), exist_ok=True)
+    else:
+        print(f"{verb} {src} -> {trg}")
+        mover(src, trg)
+
+
+def _copy_files(listing, target_dir):
+    for src in listing:
+        _move_single(src, target_dir, shutil.copy2, "Copying")
+
+
+def _link_files(listing, target_dir):
+    for src in listing:
+        _move_single(src, target_dir, os.link, "Linking")
+
+
+@timed("Moving nightly files into repo")
+def move_nightly_files(spdir, platform):
+    """Moves PyTorch files from temporary installed location to repo."""
+    # get file listing
+    source_dir = os.path.join(spdir, "torch")
+    listing = _get_listing(source_dir, platform)
+    target_dir = os.path.abspath("torch")
+    # copy / link files
+    if platform.startswith("win"):
+        _copy_files(listing, target_dir)
+    else:
+        _link_files(listing, target_dir)
+
+
+def install():
+    """Development install of PyTorch"""
+    deps, pytorch, platform = conda_solve()
```

Comment:
This nicely gets the runtime dependencies, but since we're in a new env I'm missing some essential test/dev tools. Can we add `mypy`, `pytest`, `ipython`, `sphinx`? That covers the basics I think.

Reply:
I think there should be an option to reuse an env

Reply:
I'll add both of these. In the reuse case, do you want the latest deps installed or just a straight reuse?

Reply:
I'd guess straight reuse? If you already have an env set up, it's likely recent so probably no need to download large binaries in case of a small version bump somewhere.

Reply:
Yeah - I guess my concern with straight reuse is that the dependencies are not compatible anymore with what is in the existing environment.